pkg/events/derive/hidden_kernel_module.go

```diff
@@ -62,65 +64,125 @@ func deriveHiddenKernelModulesArgs() deriveArgsFunction {
 
-		var name string
-		nameBytes, err := parse.ArgVal[[]byte](event.Args, "name")
-		if err != nil {
-			name = ""
-			// Don't fail hard, submit it without a name!
-			logger.Debugw("Failed extracting hidden module name")
-		} else {
-			// Remove the trailing terminating characters.
-			name = string(nameBytes[:bytes.IndexByte(nameBytes[:], 0)])
-		}
+		return [][]interface{}{extractFromEvent(event.Args, address)}, nil
+	}
+}
 
-		// Parse module srcversion if possible
+// InitHiddenKernelModules initializes the module components
+func InitHiddenKernelModules(modsMap *bpf.BPFMap, newModMap *bpf.BPFMap, deletedModMap *bpf.BPFMap) error {
+	allModsMap = modsMap
+	newModuleOnlyMap = newModMap
+	recentDeletedModulesMap = deletedModMap
 
-		var srcversion string
-		srcversionBytes, err := parse.ArgVal[[]byte](event.Args, "srcversion")
-		if err != nil {
-			srcversion = ""
-			// Don't fail hard, submit it without a srcversion!
-			logger.Debugw("Failed extracting hidden module srcversion")
-		} else {
-			// Remove the trailing terminating characters
-			srcversion = string(srcversionBytes[:bytes.IndexByte(srcversionBytes[:], 0)])
+	var err error
+	foundHiddenKernModsCache, err = lru.New[uint64, struct{}](2048)
+	if err != nil {
+		return err
+	}
+
+	eventsFromHistoryScan, err = lru.New[*trace.Event, struct{}](50) // If there are more hidden modules found in history scan, it'll report only the size of the LRU
+	return err
+}
+
+// handleHistoryScanFinished handles the case where the history scan finished
+func handleHistoryScanFinished(scanStatus uint64) ([][]interface{}, []error) {
+	// Address field unused in this case: use it as a flag for scan status
+	if scanStatus == 0 {
+		// Finished unsuccessfully, abort publishing events derived from this scan
+		// since the scan didn't really finish and the events might be the false positives
+		eventsFromHistoryScan.Purge()
+		return nil, nil
+	}
+
+	var res [][]interface{}
+	for {
+		e, _, ok := eventsFromHistoryScan.RemoveOldest()
+		if !ok {
+			break
 		}
 
-		addrHex := fmt.Sprintf("0x%x", address)
-		if len(addrHex) == 2 {
-			logger.Warnw("Failed converting module address to hex")
+		address, err := parse.ArgVal[uint64](e.Args, "address")
+		if err != nil {
+			return nil, []error{err}
 		}
 
-		// Add to cache not to report it multiple times
+		if _, found := foundHiddenKernModsCache.Get(address); found {
+			continue
+		}
 
 		foundHiddenKernModsCache.Add(address, struct{}{})
+		res = append(res, extractFromEvent(e.Args, address)) // Note using the event from LRU and not the event received in the derived event
+	}
 
-		return []interface{}{addrHex, name, srcversion}, nil
+	return res, nil // Send all the events
+}
+
+// extractFromEvent extract arguments from the trace.Argument
+func extractFromEvent(args []trace.Argument, address uint64) []interface{} {
+	// Parse module name if possible
+	var name string
+	nameBytes, err := parse.ArgVal[[]byte](args, "name")
```

Comment:
The logic of parsing bytes and turning them to string can be extracted to a helper function.